Make minor changes to kctrl for integration with tanzu-framework

praveenrewar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:
*kctrl*
- Use binary name to configure namespace envVar key
- Use binary name to configure kubeconfig, kubeconfig context and kubeconfig yaml envVar key
- Use status as column name instead of description in pkgr get and pkgi get to show the current reconciliation status of repository or package installation.
- Return an error in package available get if package is not present
- Add --default-values-file-output-flag to generate default values file for an available package version.

#### Which issue(s) this PR fixes:
<!--
If no issue exists for this change, please create an issue and link it here.
-->
Fixes #

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note
package available get now has a --default-values-file-output flag to generate default-values file for an available package version (with all the values commented).
```

#### Additional Notes for your reviewer:

##### Review Checklist:

- [x] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [x] Relevant tests are added or updated
- [x] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [x] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

praveenrewar committed:
Return an error if package is not present

Return an error if package metadata and package versions both are not present

praveenrewar committed:
Use binary name to configure namespace envVar key

Use binary name from pkgCmdTreeOpts to configure namespace environment variable key

praveenrewar committed:
Use binary name to configure kubeconfig envVar key

Use binary name to configure envVar key for kubeconfig, kubeconfig context and kubeconfig yaml

praveenrewar committed:
Use status as column name in pkgr get and pkgi get

Use status as column name instead of description in pkgr get and pkgi get to show the current reconciliation status of repository or package installation.

praveenrewar committed:
Fix typo in pkg install usage for positional args

Use install instead of create in pkg install when positional arguments are enabled

praveenrewar committed:
Add --default-values-file-output flag to pkga get

Add --default-values-file-output flag to package available get which lets users save the default values to a file

cppforlife commented on cli/pkg/kctrl/cmd/package/available/get.go:
```diff
@@ -65,6 +68,7 @@ func NewGetCmd(o *GetOptions, flagsFactory cmdcore.FlagsFactory) *cobra.Command
 	}
 
 	cmd.Flags().BoolVar(&o.ValuesSchema, "values-schema", false, "Values schema of the package (optional)")
+	cmd.Flags().BoolVar(&o.GenerateDefaultValues, "generate-default-values-file", false, "Generate default values from schema of the package (optional)")
```
we have a convention that flags that end with -output output something, so for this case, `default-values-file-output file.yml` would be better.

praveenrewar replied:
Just to make sure that I have understood this correctly, we want to update the flag to `default-values-file-output` and it should accept a file name where the values should be saved, right?
I have been thinking about the latter part as well (user should have the flexibility to choose file name).

cppforlife commented on cli/pkg/kctrl/cmd/package/available/generate-default-values.go:
```diff
@@ -0,0 +1,136 @@
+const Object = "object"
+
+// SchemaDefault returns a yaml byte array with values populated according to schema
+func SchemaDefault(schema []byte) ([]byte, error) {
```
lets turn this into class with all of these methods private, except one.

cppforlife commented on cli/pkg/kctrl/cmd/package/available/generate-default-values.go:
```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 VMware, Inc.
```
file names do not have dashes.

praveenrewar replied:
Hmm, I thought I was using underscore. Good catch 👍🏻

cppforlife commented on cli/pkg/kctrl/cmd/package/available/generate-default-values.go:
```diff
@@ -0,0 +1,136 @@
+	var b bytes.Buffer
+	yamlEncoder := yaml.NewEncoder(&b)
+	yamlEncoder.SetIndent(2)
+	if err := yamlEncoder.Encode(unstructured); err != nil {
```
challenge with us generating "default values" file is that if user just takes that yaml and uses it as values.yml, then user effectively says "override everything" but in general user just needs to override just a few values. i believe @jorgemoralespou proposed an idea that we should generate a file that has all the fields commented. so that user can just uncomment fields and set values to what they want. additionally it would be nice if we can include descriptions on top of values and indicate what type.

praveenrewar replied:
> we should generate a file that has all the fields commented. so that user can just uncomment fields and set values to what they want.

I really like the idea. 

> additionally it would be nice if we can include descriptions on top of values and indicate what type.

That would be great. Let me see if I can add that as part of the same PR (the title minor changes doesn't feel suitable now 😁 )

praveenrewar committed:
Update source value to exclude constraint if empty

Update source column in package repository commands to exclude constraint if empty

praveenrewar commented on cli/pkg/kctrl/cmd/package/repository/list.go:
```diff
@@ -113,7 +113,8 @@ func NewSourceValue(pkgr v1alpha1.PackageRepository) uitable.Value {
 			source = "(imgpkg) " + pkgr.Spec.Fetch.ImgpkgBundle.Image
-			if pkgr.Spec.Fetch.ImgpkgBundle.TagSelection != nil && pkgr.Spec.Fetch.ImgpkgBundle.TagSelection.Semver != nil {
+			if pkgr.Spec.Fetch.ImgpkgBundle.TagSelection != nil && pkgr.Spec.Fetch.ImgpkgBundle.TagSelection.Semver != nil &&
+				pkgr.Spec.Fetch.ImgpkgBundle.TagSelection.Semver.Constraints != "" {
```
Does this make sense or should we have a default value (something like v0.0.0+) if constraint is empty?

cppforlife replied:
package repository logic in kapp-controller is responsible for picking version if it's not provided by a human.

praveenrewar committed:
Update help and error messages in pkg commands

Bump cobrautil

praveenrewar committed:
Update default-values-file-output to have comments

Update default-values-file-output to have all the default values as comments

cppforlife commented on cli/pkg/kctrl/cmd/package/available/get.go:
```diff
@@ -269,3 +287,25 @@ func (o *GetOptions) showValuesSchema(client pkgclient.Interface, pkgName, pkgVe
+		return err
+	}
+
+	os.WriteFile(o.DefaultValuesFile, defaultValues, 0600)
```
no err check

praveenrewar replied:
Updated.
